certora/specs/LiquidityPoolMoneyFlow.spec

```diff
@@ -0,0 +1,157 @@
+// the target address whenever the value transferred is nonzero
+persistent ghost address money_recipient;
+function CVLCallHook(env e)  {
+    if (e.msg.value > 0) {
+        money_recipient = e.msg.sender;
+    }
+}
+
+// the only possible flow of validator funds is eigenpod -> etherfiNode ->
+// expectedParties where expected parties in the current contract are currently
+// bnftHolder / tnftHolder / nodeOperator / treasury. An upcoming change from @V
+// will be adjusting that mechanism slightly so that all funds flow directly to
+// the liquidity pool. But the important piece of the property to me is simply
+// that funds can't possibly be transferred to an unexpected target. This would
+// give the assurance that the worst case scenario of a logic bug is simply that
+// some validator funds could be temporarily stuck until we upgrade with a fix,
+// but user funds would never be lost.
+
+// * whichFunctionSendsETH_EtherFiNode in Basic.spec proves
+// that the only functions that move money out of EtherFiNode
+// are withdrawFunds and moveFundsToManager.
+// * withdrawFunds can only be called by EtherFiNodeManager
+// (because of a modifier) and it transfers money to addresses
+// passed by parameter
+// * EtherFiNodesManager calls EtherFiNode.withdraw in
+// distributePayouts which is internal and called by: fullWithdraw,
+// partialWithdraw
+// * whichFunctionSendsETHFromNode_NodesManager also
+// proves that fullWithdraw / partialWithdraw
+// are the only functions which cause 
+
+rule money_flow_from_node_full_withdraw {
+    uint256 validatorId;
+    env e;
+
+    // initialize ghost to currentContract
+    require money_recipient == currentContract; 
+
+    // expected party addresses
+    address treasury = currentContract.treasuryContract;
+    address nodeOperator = auctionManager.getBidOwner(e, validatorId);
+    address bnftHolder = currentContract.bnft.ownerOf(e, validatorId);
+    address tnftHolder = currentContract.tnft.ownerOf(e, validatorId);
+
+    fullWithdraw(e, validatorId);
+
+    // If money was moved during this call it goes
+    // to one of the expected recipients. The
+    // case where money_recipient is currentContract
+    // models the case where nonzero money is not moved.
+    // temporarily made satisfy to check that this is a real result
+    // with the routing that I expect
+    assert money_recipient == currentContract ||
+        money_recipient == treasury ||
+        money_recipient == nodeOperator ||
+        money_recipient == bnftHolder ||
+        money_recipient == tnftHolder; 
+
+    // The following encodes that all of these recipient
+    // cases are reachable (and ensures that the above
+    // assertion isn't just trivially stuck in the initial state)
+
+    // This unconstrained ghost on which we branch is
+    // just here to encode nondeterminstic choice
+    mathint nondeterministic_choice;
+    if (nondeterministic_choice == 0) {
```

Comment:
This if -else block might miss a case where all of the recipients addresses are distinct.
If all addresses are the same, you would still be able to satisfy every case here.

Reply:
In the previous version of this rule they were separated by different if-else cases, so they couldn't assume each other.

Reply:
I'm still in favor of the new satisfy statement you've added. It guarantees the most general case is satisfiable.

Reply:
> In the previous version of this rule they were separated by different if-else cases, so they couldn't assume each other.

Yes, I am aware this was exactly the point.